### PR TITLE
CRI: Add Windows Podsandbox Stats

### PR DESCRIPTION
**Pull request by jsturtevant**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Adds fields to the cri api for Windows to collect metrics from the Container runtime.

Part of https://github.com/kubernetes/enhancements/issues/2371

Follow up to the initial commit that added Linux stats https://github.com/kubernetes/kubernetes/pull/102789

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/2371
```

/sig node
/sig windows

/cc @dcantah @haircommander @bobbypage @marosset

### CODE REVIEW

**Comment by jsturtevant** [comment]
The hyperv stats collected for the VM are https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/stats/stats.proto#L47-L70

**Comment by dcantah** [comment]
> The hyperv stats collected for the VM are https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/stats/stats.proto#L47-L70

Sorry, thought I put these in the post 🥲

**Comment by jsturtevant** [comment]
I went through both types of stats and compared to what we have now and what would be available in hyperv.  We don't use the whole set of metrics for existing containers.   CPU stats look fine, only one memory field is used currently (as @dcantah pointed out) and network is missing two fields but I think we can use 

We have a few options: 

- stay with currently exposed metrics regardless of if we use them or they make sense for windows
- create new memory and network structure that map to what we want for pod sandbox calls
- create new memory and network structure that map to what we want for pod sandbox calls and update the container stats calls to use the new structs.  No exactly sure how this would work.

Some info:


**cpu usage** - https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/stats/stats.proto			

| field                      | type   | process isolated field | hyperv filed     | notes                                                                                                                                                                                                                |
| -------------------------- | ------ | ---------------------- | ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| timestamp                  | int64  | ✅                      | ✅                |                                                                                                                                                                                                                      |
| usage\_core\_nano\_seconds | uint64 | TotalRuntimeNS         | TotalRuntimeNS   | // Cumulative CPU usage (sum across all cores) since object creation.                                                                                                                                                |
| usage\_nano\_cores         | uint64 | calculated value       | calculated value | calculated value done runtime (containerd or kubelet)<br><br> // Total CPU usage (sum of all cores) averaged over the sample window.  <br> // The "core" unit can be interpreted as CPU core-nanoseconds per second. |

**Memory usage** - https://github.com/microsoft/hcsshim/blob/master/cmd/containerd-shim-runhcs-v1/stats/stats.proto			

| field               | type   | process isolated field                      | hyperv filed         | notes                                                                                                                                                               |
| ------------------- | ------ | ------------------------------------------- | -------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| timestamp           | int64  | ✅                                           | ✅                    |                                                                                                                                                                     |
| working\_set\_bytes | uint64 | memory\_usage\_private\_working\_set\_bytes | working\_set\_bytes? | // The amount of working set memory in bytes.<br><br>Is hyper-v working\_set\_bytes same as private\_working\_set\_bytes                                            |
| available\_bytes    | uint64 |                                             | available\_memory    |   Feels like we should be able to return this. It is the limit<br><br>  // Available memory for use. This is defined as the memory limit - workingSetBytes.<br><br> |
| usage\_bytes        | uint65 | ?                                           | ?                    | // Total memory in use. This includes all memory regardless of when it was accessed.                                                                                |
| rss\_bytes          | uint66 | n/a                                         | n/a                  | windows doesn't do rss                                                                                                                                              |
| page\_faults        | uint67 |                                             |                      | not reported. Is possible to report?                                                                                                                                |
| major\_page\_faults | uint68 |                                             |                      | not reported. Is possible to report?                                                                                                                                |


**Process isolated** also has

- memory_usage_commit_bytes
- memory_usage_commit_peak_bytes 

**Hyperv** also has 

- virtual_node_count
- available_memory_buffer
- reserved_memory 
- assigned_memory 
- slp_active 
- balancing_enabled
- dm_operation_in_progress 

These are not used but I am not sure it makes a tone of sense to include them all as most would be very specific and could be collected by specialized tools on a as needed basis.

**Network stats**
| field      | type   | process isolated field | hyperv filed | notes                                                                                                                                                 |
| ---------- | ------ | ---------------------- | ------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
| timestamp  | int64  | ✅                      | ✅            |                                                                                                                                                       |
|  name      | uint64 |      EndpointID                  |              | same values can be used?                                                                                                                                                         |
| rx\_bytes  | uint65 | BytesReceived          |              |   same values can be used?                                                                                                                                                    |
| rx\_errors | uint66 |                        |              | can we use DroppedPacketsIncoming? https://github.com/microsoft/hcsshim/blob/949e46a1260a6aca39c1b813a1ead2344ffe6199/internal/hns/hnsendpoint.go#L65 |
| tx\_bytes  | uint67 | BytesSent              |              |  same values can be used?                                                                                                                                                        |
| tx\_errors | uint68 |                        |              | DroppedPacketsOutgoing                                                                                                                                |

**Comment by marosset** [comment]
> We have a few options:

> - stay with currently exposed metrics regardless of if we use them or they make sense for windows
> - create new memory and network structure that map to what we want for pod sandbox calls
> - create new memory and network structure that map to what we want for pod sandbox calls and update the container stats calls to use the new structs. No exactly sure how this would work.

I think we should make new memory and network structures that map to what metrics we can collect on Windows.
I also think should add structures specific to either process isolated or hyper-v isolated sandboxes instead of just having all of the fields under a single struct and leaving them blank.

Something like

```protobuf
message WindowsProcessIsolatedMemoryUsage {
	unit64 timestamp = 1;
	unit64 memory_usage_commit_bytes = 2;
	unit64 memory_usage_commit_peak_bytes = 3;
	etc...
}

message WindowsHyperVIsolatedMemoryUsage {
	unit64 timestamp = 1;
	unit64 reservered_memroy_bytes = 2;
	unit64 assigned_memory_bytes = 3;
	bool slp_active = 4;
	bool balancing_enabled = 5;
	etc....
}

message WindowsMemoryUsage {
	WIndowsProcessIsolatedMemoryUsage processIsolatedMemory = 1;
	WindowsHyperVIsolatedMemoryUsage hyperVIsolatedMemory = 2;
}

```

We're not going to have many chances to revisit this so I think we should design it as we see best.

If we can update the container stats calls too that would be great!

**Comment by marosset** [comment]
> My understanding was sandbox was an abstraction and we shouldn't necessarily have different structures here since we don't want to leak the implementation detail that the sandbox is a process isolated container, hyperv or something else.

I'm curious what node folks think here.
I know there is a desired to keep the sandbox abstraction away from workloads (pod specs, events, etc) but having these metrics is essential for cluster operators who want to closely monitor performance of their nodes.
With Windows the sandbox types are tightly coupled with the operating system and do act very differently so maybe this is OK?

**Comment by jsturtevant** [comment]
> I know there is a desired to keep the sandbox abstraction away from workloads (pod specs, events, etc) but having these metrics is essential for cluster operators who want to closely monitor performance of their nodes.
> With Windows the sandbox types are tightly coupled with the operating system and do act very differently so maybe this is OK?

@dchen1107 @derekwaynecarr @haircommander to chime in on this

**Comment by bobbypage** [comment]
> I know there is a desired to keep the sandbox abstraction away from workloads (pod specs, events, etc)

+1

If we can reuse the sandbox stats protos across windows and linux and map them accordingly I think that would be ideal instead of introducing new sets of CRI stats just for windows. Is there something that is missing in the current `CpuUsage` `MemoryUsage`, `NetworkUsage`, etc messages?

**Comment by harshitasao** [comment]
Hi @jsturtevant, k8s 1.25 Bug Triage Shadow here, just checking in to see if this is still on track for k8s 1.25?
Thank you.

**Comment by jsturtevant** [comment]
> Hi @jsturtevant, k8s 1.25 Bug Triage Shadow here, just checking in to see if this is still on track for k8s 1.25?

Yes would like to have this in 1.25 so we can be ready for when this Cri-only stats is turned on by default in a Beta.  It also unblocks getting this into containerd 1.7: https://github.com/containerd/containerd/pull/7099. I've added it to sig-windows and sig-node agendas.

**Comment by marosset** [comment]
> > If we can reuse the sandbox stats protos across windows and linux and map them accordingly I think that would be ideal instead of introducing new sets of CRI stats just for windows Is there something that is missing in the current CpuUsage MemoryUsage, NetworkUsage, etc messages?
> 
> @bobbypage there was already a `WindowsPodSandboxStats` struct that was added but left blank in the original PR: #102789. Since we have this field we have the ability to define stats that make sense for windows.
> 
> There are things like `rss_bytes` which don't have an equivalent value in linux (see [#110754 (comment)](https://github.com/kubernetes/kubernetes/pull/110754#issuecomment-1175552676) for more details other fields too). So we could keep doing it the same way by re-using the existing `MemoryUsage` stats and leave fields like that blank or take this opportunity to make stats (atleast for podsandbox) that make sense of Windows as @marosset suggested.
> 
> I agree with Mark we have an opportunity to design this specifically for Windows since the field `WindowsPodSandboxStats` is already there. If we did that, I lean toward keep the stats more generic vs exposing specific metrics for the various types of runtimes under the hood.

I've thought about this some more and think that if we need to expose metrics for the various types of runtimes it would be more appropriate to have the container runtime expose those metrics directly. 
I do think there still is some benefit to having dedicated Windows structs for stats returned over CRI-API but agree they should be kept generic.

**Comment by jsturtevant** [comment]
I think the remaining questions are:

- do we re-use the  `CpuUsage` and `ProcessUsage` fields since we can meet the requirements?
- What fields to include in `WindowsMemoryUsage`? I propose we start small and add as we learn more
- Do we modify the ` repeated NetworkInterfaceUsage` to include `WindowsMemoryUsage` and other modifications and make changes to `containerstats` call

**Comment by harshitasao** [comment]
Hello @jsturtevant, with the code freeze approaching, is this still on track for v1.25?
Thanks!

**Comment by jsturtevant** [comment]
We are hashing out the details in https://github.com/kubernetes/enhancements/pull/3439.  If that gets merged then the changes here should be pretty quick to review

**Comment by jsturtevant** [comment]
updated based on feedback in https://github.com/kubernetes/enhancements/pull/3439.

**Comment by haircommander** [comment]
when do we intend on using the top-level fields in the kubelet? should we fall back to the linux fields if top-level are not specified for a release or two?

**Comment by jsturtevant** [comment]
> when do we intend on using the top-level fields in the kubelet? should we fall back to the linux fields if top-level are not specified for a release or two?

Yes this makes sense to me, we would need to get the changes for Linux into a containerd release.

**Comment by cici37** [comment]
Hello 👋, 1.25 Release Lead here.

Unfortunately, this enhancement did not meet the [code freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze) criteria because there are still unmerged k/k code PRs.

If you still wish to progress this enhancement in v1.25, please file an [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md). Thank you so much!

/milestone clear

**Comment by jsturtevant** [comment]
We will be moving forward with Windows specific fields https://github.com/kubernetes/enhancements/pull/3439

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 1]
```diff
@@ -639,7 +639,6 @@ message PodSandboxAttributes {
 }
 
 // PodSandboxStats provides the resource usage statistics for a pod.
-// The linux or windows field will be populated depending on the platform.
```
Comment (haircommander) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:642]: This comment still seems correct to me

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto [hunk 2]
```diff
@@ -658,7 +657,16 @@ message LinuxPodSandboxStats {
 
 // WindowsPodSandboxStats provides the resource usage statistics for a pod sandbox on windows
 message WindowsPodSandboxStats {
-    // TODO: Add stats relevant to windows.
+    // CPU usage gathered for the pod sandbox.
```
Comment (haircommander) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:660]: how close are we to dropping v1alpha2 @mrunalp @mikebrow ? I feel like we don't need to add this here

**Comment by haircommander** [comment]
a couple of notes, otherwise LGTM

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 3]
```diff
@@ -678,6 +686,16 @@ message NetworkUsage {
     repeated NetworkInterfaceUsage interfaces = 3;
 }
 
+// WindowsNetworkUsage contains data about network resources specific to Windows.
+message WindowsNetworkUsage {
+    // The time at which these stats were updated.
```
Comment (bobbypage) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:691]: Nit: Keep the timestamp comment consistent 

```
Timestamp in nanoseconds at which the information were collected. Must be > 0.
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 4]
```diff
@@ -700,6 +732,14 @@ message ProcessUsage {
     UInt64Value process_count = 2;
 }
 
+// WindowsProcessUsage are stats pertaining to processes specific to Windows.
+message WindowsProcessUsage {
+    // The time at which these stats were updated.
```
Comment (bobbypage) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:737]: Ditto

```
Timestamp in nanoseconds at which the information were collected. Must be > 0.
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto [hunk 5]
```diff
@@ -1554,6 +1629,18 @@ message MemoryUsage {
...
+// WindowsMemoryUsage provides the memory usage information specific to Windows
+message WindowsMemoryUsage {
+    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
+    int64 timestamp = 1;
+    // The amount of working set memory in bytes.
+    UInt64Value working_set_bytes = 2;
+    // Available memory for use. This is defined as the memory limit - workingSetBytes.
+    UInt64Value available_bytes = 3;
+    // Cumulative number of page faults.
+    UInt64Value page_faults = 6;
```
Comment (bobbypage) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1641]: why does this jump from proto tag=3 to tag=6 ?

**Comment by bobbypage** [comment]
couple of small comments, also curious if we need the v1alpha2 changes or can stick with just v1?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 3 above)
Comment (jsturtevant) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:691]: This was from linux network usage comment above, I'll updated both to be consistent.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 5 above)
Comment (jsturtevant) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:1641]: typo, good catch!

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 2 above)
Comment (jsturtevant) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: I dropped the v1alpha2 for now.  Let me know if I should add it back

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto (hunk 1 above)
Comment (jsturtevant) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto]: added back!

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto (hunk 2 above)
Comment (jsturtevant) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto]: Looks like there are tests that validate v1alpha2 https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/110754/pull-kubernetes-unit/1583552443948470272

**Review by marosset — APPROVED** [review]
/lgtm

**Comment by jsturtevant** [comment]
updated based on the feedback @bobbypage @haircommander

**Comment by haircommander** [comment]
/lgtm

thanks!

**Comment by jsturtevant** [comment]
/assign @dchen1107 @derekwaynecarr @mikebrow

Implementing the changes in https://github.com/kubernetes/enhancements/pull/3439

**Review by mrunalp — APPROVED** [review]

**Pull request merged**